engine/src/parse_file.rs

```diff
@@ -222,36 +273,79 @@ pub trait CppBuildable {
 impl ParsedFile {
     /// Get all the autocxxes in this parsed file.
     pub fn get_rs_buildables(&self) -> impl Iterator<Item = &IncludeCppEngine> {
-        self.0.iter().filter_map(|s| match s {
-            Segment::Autocxx(includecpp) => Some(includecpp),
-            _ => None,
-        })
+        fn do_get_rs_buildables(
+            segments: &Vec<Segment>,
+        ) -> impl Iterator<Item = &IncludeCppEngine> {
+            segments
+                .iter()
+                .flat_map(|s| -> Box<dyn Iterator<Item = &IncludeCppEngine>> {
```

Comment:
Do you think you could abstract this repetition out into some function which returns an iterator of all segments that are not `Segment::Mod`? Then implement all three of these functions in terms of that iterator?

Reply:
It ends up being 2 functions to split out the `iter_mut` (or pass even more closures around), and getting all the generic types and lifetimes to work out seems really ugly. I managed to figure out something that works, but it was mostly as an exercise in me learning how to work with lifetimes and trait objects and type bounds in the end, I don't think it's a good idea. I might be missing a neater way to write the types, but I think the way the lifetimes have to be managed is fundamentally complicated.

Ignoring the lifetime and type bounds, it still ends up as about the same number of lines of code, and IMHO they're more complicated lines. The logic is split out from the infrastructure, but I'm not sure it's better overall.

I pushed what I ended up with as 31d000cb7dee8c15deccfe09c0afbbdce60bf5c5, do you see a better way to do it / do you want all of that?

Reply:
Yeah that's horrid :) I think I might be able to do something a little nicer, though I'm probably deluding myself. I'll have a crack at it when I get a chance.

Reply:
OK, I had in mind something like this: https://github.com/google/autocxx/pull/new/segment-iterator-attempts. It's an iterator which contains a stack of nested mods to explore.

It has no test code so therefore probably has bugs :) Also, I didn't do a `mut` version.

This doesn't successfully shorten the code either.

On the whole I have a marginal preference for my approach, even though it's longer, because we're not repeating the logic related to recursing through the tree. However, I also just don't care very much either way, and as this would end up being longer (after the `mut` version) I'm quite happy with your approach. Up to you.

Reply:
I think I prefer mine, because I find it easier to work with despite the repetition.

Reply:
Fine with me.

Reply:
Fine with me